include/ruby/st.h

```diff
@@ -79,6 +79,7 @@ struct st_table_entry; /* defined in st.c */
 struct st_table {
     /* Cached features of the table -- see st.c for more details.  */
     unsigned char entry_power, bin_power, size_ind;
+    uint8_t shared;
```

Comment:
You can likely use `unsigned int` here, because the compiler will pad that single byte to 4 or 8 bytes for alignment purposes.

And then use `UINT_MAX` to check for overflow.

Reply:
Actually, seems like I'm wrong. On arm64 at least `sizeof(st_table)` jump from `56` to `64` when changing this for `unsigned int`.

Reply:
Other potential tradeoffs here:

  - `st_table` is used by other things than `Hash`, so it might be seen as non-optimal to add this field here. It might be preferable for `hash.c` to have it's own struct with `st_table` as a member.
  - For (system) CoW performance, it would be best if duping a frozen Hash wouldn't mutate that original `st_table`. See https://github.com/ruby/ruby/pull/6171 for instance.

Reply:
Also for `Array` and `String`, CoW is done at the Ruby Object level, here you do it one level lower, which might be fine, but probably need some argument for the inconsistency.

Reply:
There are no spare flags on the Hash object so we can't implement the same model for CoW.

Reply:
I know but it's very easy to reclaim 2 bits from the iterlevel mask.

Right now it allows for 127 iterlevel before falling back to storing it in ivar, if you take two bits out, that puts us at `31` iterlevel before using an ivar.

No reasonable code goes over 2 or 3 iterlevel, and the unreasonable or rare code that does will gracefully degrade.